test/Base_Tests/src/Data/Decimal_Spec.enso

```diff
@@ -492,6 +500,76 @@ add_specs suite_builder =
+            c = Decimal.new "12000" (Math_Context.new 3)
+            huge_a = Decimal.new "3.4E320"
+            huge_b = Decimal.new "3.400E320"
+            huge_c = Decimal.new "3.4000000E320"
+            huge_integer = 34 * (10 ^ 319)
+
+            a.to_integer . should_equal 12000
+            b.to_integer . should_equal 12000
+            c.to_integer . should_equal 12000
+            huge_a.to_integer . should_equal huge_integer
+            huge_b.to_integer . should_equal huge_integer
+            huge_c.to_integer . should_equal huge_integer
+
+            (-a).to_integer . should_equal -12000
+            (-b).to_integer . should_equal -12000
+            (-c).to_integer . should_equal -12000
+            (-huge_a).to_integer . should_equal -huge_integer
+            (-huge_b).to_integer . should_equal -huge_integer
+            (-huge_c).to_integer . should_equal -huge_integer
+
+        group_builder.specify "should convert correctly to Float" <|
+            a = Decimal.new "12000"
+            b = Decimal.new "12000" (Math_Context.new 2)
+            c = Decimal.new "12000" (Math_Context.new 3)
+            huge_a = Decimal.new "3.4E300"
+            huge_b = Decimal.new "3.400E300"
+            huge_c = Decimal.new "3.4000000E300"
+            huge_float = 3.4 * (10 ^ 300)
+
+            a.to_float . should_equal 12000.0
+            b.to_float . should_equal 12000.0
+            c.to_float . should_equal 12000.0
+            huge_a.to_float . should_equal huge_float
+            huge_b.to_float . should_equal huge_float
+            huge_c.to_float . should_equal huge_float
+
+            (-a).to_float . should_equal -12000
+            (-b).to_float . should_equal -12000
+            (-c).to_float . should_equal -12000
+            (-huge_a).to_float . should_equal -huge_float
+            (-huge_b).to_float . should_equal -huge_float
+            (-huge_c).to_float . should_equal -huge_float
+
+        group_builder.specify "should convert correctly to Integer and Float using implicit conversions" <|
+            i = Integer.from (Decimal.new "5634")
+            i.should_be_a Integer
+            i.should_equal 5634
+
+            f = Float.from (Decimal.new "56.34")
+            f.should_be_a Float
+            f.should_equal 56.34
+
+        group_builder.specify "Decimal.to_float should compare correctly with the original Decimal" <|
+            huge_a = Decimal.new "3.4E300"
+            huge_a_float = Float.from huge_a
+
+            (huge_a_float == huge_a) . should_be_true # succeeds
+            (huge_a == huge_a_float) . should_be_true # fails
+
+        group_builder.specify "Decimal.to_integer should compare correctly with the original Decimal" pending="Decimal.to_integer comparison" <|
+            huge_a = Decimal.new "3.4E320"
+            huge_a_int = Integer.from huge_a
+
+            (huge_a_int == huge_a) . should_be_true # succeeds
+            (huge_a == huge_a_int) . should_be_true # fails
```

Comment:
@JaroslavTulach I suspect that I am improperly returning a value from `.to_integer` here, resulting in this failure.

Reply:
Isn't this the missing `ToEnsoNumberNode` component in `HostValueToEnsoNode`?

Reply:
`huge_a_int` shall be converted to `Decimal` and the _decimal comparator_ shall be invoked to compute `hash` and `compare`. If that's not happening, only debugger can tell us what's going on.

Reply:
The selected specialization is [EqualsSimpleNode.equalsDifferent](https://github.com/enso-org/enso/blob/55a47dac20ec679fb449cbdebef8aa8618dc71f7/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java#L267). It is true that the _"primitiveness"_ of both values is different:
```
JavaObject[340000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 (java.math.BigInteger)]
isPrimitive(other, interop)
true
isPrimitive(self, interop)
false
self
Value JavaObject[3.4E+320 (java.math.BigDecimal)]
```
but there is a **conversion between one and another** - e.g. we should be continuing to [EqualsNode.findConversions](https://github.com/enso-org/enso/blob/55a47dac20ec679fb449cbdebef8aa8618dc71f7/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java#L151). Right, @Akirathan?

I can take a look when I am back from vacation.

Reply:
Trying the alternative approach of moving `TypeOfNode` behavior to `TypesLibrary` [here](https://github.com/enso-org/enso/compare/wip/gmt/8982-conversion-int-float...JaroslavTulach:enso:wip/jtulach/UseTypesLibrary) - not yet fully working, but also not fully broken.

Reply:
> Trying the alternative approach of moving TypeOfNode behavior to TypesLibrary [here](https://github.com/enso-org/enso/compare/wip/gmt/8982-conversion-int-float...JaroslavTulach:enso:wip/jtulach/UseTypesLibrary) - not yet fully working, but also not fully broken.

@JaroslavTulach Why not the other way around? That is, make the `TypeOfNode` the default way to fetch the type. I thought that using nodes is more light-weight than using libraries.

Reply:
@JaroslavTulach This patch does fix the failing test case.

Reply:
Bug for this: #9547